File: navi_and_math.py
from itertools import combinations
from numpy import product
from operator import mul,add
import sys
from decimal import Decimal
from math import fmod
T=int(sys.stdin.readline())
M=10**9+7
for i in range(T):
    N=int(sys.stdin.readline())
    L=map(Decimal, sys.stdin.readline().split())
    if len(L)==2: 
        A=L
        res=(product(L)/sum(L))%M
    else: 
        A=sum([map(list,combinations(L,x)) for x in range(2,N+1)],[])[1:]
        R=map(lambda x:fmod(product(x)/sum(x),M),A)
        print(R)
        res=max(R)
    print("Case #"+str(i+1)+": "+"%.f"%(res))
    T-=1

File: scala_problems_competitive/Navi_And_Math.scala
package com.hackerearth
import scala.math.BigInt
object NaviAndMath {   
  def main(args: Array[String]): Unit = {
      val lines = io.Source.stdin.getLines()
      val M:BigInt=BigInt(1000000007)
      val T=lines.next().toInt
      for(i <- 1 to T) {
        val N=lines.next().toInt        
        val l = lines.next().split(" ").map(BigInt(_)).toList
        val r: IndexedSeq[BigInt]=(for{x<-2 to N 
          e=l.view.combinations(x).map(x => x.product/x.sum)                   
        } yield e).flatten.sortWith(_<_)
        val res= r.partition(x => x<=M)._1.last%M
        println("Case #"+i.toString()+": "+res.toString())        
      }      
  }
}